File: python/src/coding_interview/infrastructure/repository/market_price_repository_impl.py
from __future__ import annotations

from decimal import Decimal

from coding_interview.domain.constants import INITIAL_PRICES
from coding_interview.domain.stock_symbol import StockSymbol


class MarketPriceRepositoryImpl:
    def __init__(self) -> None:
        self._prices: dict[StockSymbol, Decimal] = dict(INITIAL_PRICES)

    def all(self) -> dict[StockSymbol, Decimal]:
        return dict(self._prices)

    def update(self, prices: dict[StockSymbol, Decimal]) -> None:
        self._prices = dict(prices)

File: ruby/lib/coding_interview/infrastructure/repository/market_price_repository_impl.rb
require_relative "../../application/repository/market_price_repository"
require_relative "../../domain/app_constants"

module CodingInterview
  module Infrastructure
    module Repository
      class MarketPriceRepositoryImpl < Application::Repository::MarketPriceRepository
        def initialize
          @prices = Domain::AppConstants::INITIAL_PRICES.dup
          @mutex = Mutex.new
        end

        def all
          @mutex.synchronize { @prices.dup }
        end

        def update(prices)
          @mutex.synchronize { @prices = prices.dup }
          nil
        end
      end
    end
  end
end
